nixos/modules/services/continuous-integration/gitea-actions-runner.nix

```diff
@@ -0,0 +1,176 @@
+{ config
+, lib
+, pkgs
+, utils
+, ...
+}:
+
+let
+  inherit (lib)
+    any
+    attrValues
+    concatStringsSep
+    escapeShellArg
+    hasInfix
+    optionalAttrs
+    optionals
+    literalExpression
+    mapAttrs'
+    mkEnableOption
+    mkOption
+    mkPackageOptionMD
+    mkIf
+    nameValuePair
+    types
+  ;
+
+  inherit (utils)
+    escapeSystemdPath
+  ;
+
+  cfg = config.services.gitea-actions-runner;
+
+  usesContainers = any (instance:
+    # the default label set includes ubuntu containers
+    instance.labels == [] ||
+    # explicit container usage
+    (any (label: hasInfix label "docker://") instance.labels)
+  ) cfg.instances;
+
+  usesDocker = usesContainers && config.virtualisation.docker.enable;
+  usesPodman = usesContainers && config.virtualisation.podman.enable;
+in
+{
+  meta.maintainers = with lib.maintainers; [
+    hexa
+  ];
+
+  options.services.gitea-actions-runner = with types; {
+    package = mkPackageOptionMD pkgs "gitea-actions-runner" { };
+
+    instances = mkOption {
+      default = {};
+      description = lib.mdDoc ''
+        Gitea Actions Runner instances.
+      '';
+      type = attrsOf (submodule {
+        options = {
+          enable = mkEnableOption (lib.mdDoc "Gitea Actions Runner instance");
+
+          name = mkOption {
+            type = str;
+            example = literalExpression "config.networking.hostName";
+            description = lib.mdDoc ''
+              The name identifying the runner instance towards the Gitea/Forgejo instance.
+            '';
+          };
+
+          url = mkOption {
+            type = str;
+            example = "https://forge.example.com";
+            description = lib.mdDoc ''
+              Base URL of your Gitea/Forgejo instance.
+            '';
+          };
+
+          token = mkOption {
+            type = nullOr str;
+            default = null;
+            description = lib.mdDoc ''
+              Plain token to register at the configured Gitea/Forgejo instance.
+            '';
+          };
+
+          tokenFile = mkOption {
+            type = nullOr path;
+            default = null;
+            description = lib.mdDoc ''
+              Path to an environment file, containing the `TOKEN` environment
+              variable, that holds a token to register at the configured
+              Gitea/Forgejo instance.
+            '';
+          };
+
+          labels = mkOption {
+            type = listOf str;
+            example = literalExpression ''
+              [
+                "host:host"
+                "nixos:docker://nixos/nix:latest"
+                "ubuntu-rolling:docker://ubuntu:rolling"
+                "ubuntu-lts:docker//ubuntu:latest"
+              ]
+            '';
+            description = lib.mdDoc ''
+              Labels used to map jobs to their runtime environment.
+            '';
+            apply = concatStringsSep ",";
+          };
```

Comment:
Default copied from https://gitea.com/gitea/act_runner/src/commit/b49834185777fffed9e0b4d7c3b08ad82fcdbcfa/internal/app/cmd/register.go#L93-L98

I think this is much better, as it is much clearer, allows for stable default in the stable channel and makes assertions easier.

The downside is, that we might have to bump this value on unstable from time to time when `pkgs.gitea-actions-runner` updates.

Also, as far as I am aware, all docker-based runner based docker images need to have node in them to run the actual actions.

Will need to double check later.

```suggestion
          labels = mkOption {
            type = listOf str;
            example = [
              "host:host"
              "nixos:docker://nixos/nix:latest"
              "ubuntu-22.04:docker://node:16-bullseye"
            ];
            default = [
              "ubuntu-latest:docker://node:16-bullseye"
              "ubuntu-22.04:docker://node:16-bullseye"
              "ubuntu-20.04:docker://node:16-bullseye"
              "ubuntu-18.04:docker://node:16-buster"
            ];
            description = lib.mdDoc ''
              Labels used to map jobs to their runtime environment.
            '';
            apply = concatStringsSep ",";
          };
```

Reply:
> The downside is, that we might have to bump this value on unstable from time to time when pkgs.gitea-actions-runner updates.

Not a fan. Would really appreciate not having to migrate versions all the time. That's why my example contained only rolling or unversioned labels.